packages/core/saved-objects/core-saved-objects-migration-server-internal/src/zdt/model/stages/documents_update_init.ts

```diff
@@ -8,39 +8,88 @@
   'DOCUMENTS_UPDATE_INIT',
-  'SET_DOC_MIGRATION_STARTED'
+  'SET_DOC_MIGRATION_STARTED' | 'DONE' | 'FATAL'
 > = (state, res, context) => {
   if (Either.isLeft(res)) {
     throwBadResponse(state, res as never);
   }
 
-  const excludeFilterHooks = Object.fromEntries(
-    context.types
-      .map((name) => context.typeRegistry.getType(name)!)
-      .filter((type) => !!type.excludeOnUpgrade)
-      .map((type) => [type.name, type.excludeOnUpgrade!])
-  );
-
   const types = context.types.map((type) => context.typeRegistry.getType(type)!);
-  const outdatedDocumentsQuery = getOutdatedDocumentsQuery({ types });
+  const logs: MigrationLog[] = [...state.logs];
+  const versionCheck = checkVersionCompatibility({
+    mappings: state.previousMappings,
+    types,
+    source: 'docVersions',
+    deletedTypes: context.deletedTypes,
+  });
 
-  const transformRawDocs = createDocumentTransformFn({
-    serializer: context.serializer,
-    documentMigrator: context.documentMigrator,
+  logs.push({
+    level: 'info',
+    message: `DOCUMENTS_UPDATE_INIT: doc version check result: ${versionCheck.status}`,
   });
 
-  return {
-    ...state,
-    excludeOnUpgradeQuery: excludeUnusedTypesQuery,
-    excludeFromUpgradeFilterHooks: excludeFilterHooks,
-    outdatedDocumentsQuery,
-    transformRawDocs,
-    controlState: 'SET_DOC_MIGRATION_STARTED',
-  };
+  switch (versionCheck.status) {
+    // app version is greater than the index mapping version.
+    // scenario of an upgrade: we need to run the document migration.
+    case 'greater':
+      const excludeFilterHooks = Object.fromEntries(
+        context.types
+          .map((name) => context.typeRegistry.getType(name)!)
+          .filter((type) => !!type.excludeOnUpgrade)
+          .map((type) => [type.name, type.excludeOnUpgrade!])
+      );
+      const outdatedDocumentsQuery = getOutdatedDocumentsQuery({ types });
+      const transformRawDocs = createDocumentTransformFn({
+        serializer: context.serializer,
+        documentMigrator: context.documentMigrator,
+      });
+      return {
+        ...state,
+        logs,
+        excludeOnUpgradeQuery: excludeUnusedTypesQuery,
+        excludeFromUpgradeFilterHooks: excludeFilterHooks,
+        outdatedDocumentsQuery,
+        transformRawDocs,
+        controlState: 'SET_DOC_MIGRATION_STARTED',
+      };
+    // app version and index mapping version are the same.
+    // either application upgrade without model change, or a simple reboot on the same version.
+    // There's nothing to do here, as documents are already at the same version.
+    case 'equal':
+      return {
+        ...state,
+        logs,
+        controlState: 'DONE',
+      };
+    // app version is lower than the index mapping version.
+    // Should only occur in case of rollback.
+    // For now, we don't do anything.
+    case 'lesser':
+      return {
+        ...state,
+        logs,
+        controlState: 'DONE',
+      };
```

Comment:
For now, we just perform a noop for downgrades and go straight to DONE. It may change in the future, depending on how we want to manage that, but it probably makes sense to just have a defined behavior for now, to be able to add tests around it.